Firestore/core/src/firebase/firestore/model/document_key.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_KEY_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_KEY_H_
+
+#include <initializer_list>
+#include <memory>
+#include <string>
+
+#include "absl/strings/string_view.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+class ResourcePath;
```

Comment:
I'm not sure this build-time optimization is worthwhile. It forces the entire implementation out into the .cc file, where this would otherwise be a very lightweight wrapper around a `ResourcePath`.

I don't feel strongly that these things should be defined inline and perhaps this is beneficial because it forces all these things to be defined out-of-line.

Reply:
Reverted back.